src/main/scala/rps/Game.scala

```diff
@@ -1,30 +1,40 @@
 package rps
 
+import RPSMoves._
+
 object Game {
   def play(): Unit = {
 
     /* Ask for user input */
-    val userMove = readLine(s"""
+    val userMoveString = readLine(s"""
     |What's your move?"
-    |0: ${returnMoveIcon("0")}
-    |1: ${returnMoveIcon("1")}
-    |2: ${returnMoveIcon("2")}
+    |0: ${ROCK.toString}
+    |1: ${PAPER.toString}
+    |2: ${SCISSORS.toString}
     |
     |> """.stripMargin)
 
-    val computerMove = generateCPUMove();
+    val userMove = RPSMoves.factoryMove(userMoveString)
+
+    if (userMove == None) {
+      println(s"Mmm looks like your move was not legal... 🤔")
+    } else {
 
-    println(s"Your move was:    ${returnMoveIcon(userMove)}")
-    println(s"The CPU move was: ${returnMoveIcon(computerMove)}")
+    val computerMoveString = generateCPUMove()
+    val computerMove = RPSMoves.factoryMove(computerMoveString)
+
+    println(s"Your move was:    ${userMove.get.toString}")
+    println(s"The CPU move was: ${computerMove.get.toString}")
 
     /* Compute the outcome */
-    computeGameOutcome(userMove, computerMove)
+    computeGameOutcome(userMove.get, computerMove.get)
+    }
   }
 
   private def generateCPUMove(): String =
```

Comment:
`generateCPUMove` could return directly an `RPSMove`

Reply:
For lazyness I didn't change that at all, my bad. :3 

But let's pretend generateCPUMove was public and people was using it (thus I could not change its signature so easily).

Could I change something in such a way generateCPUMove was *simultaneously* returning a string or a RPSMove object? (in this way, the public method would not break the existing calls but would be enhanced anyway with the new support for RPSMove datatype?)

I know this question sounds strange (in C is impossible/not advisable), but maybe some scala features allow this (or even better, it is a well known design pattern).

Reply:
What you're suggesting is overloading a method by the return type, which is not possible in Scala (and in any other language I'm aware of, FWIW, since it's super hard to typecheck and not so useful).

So, no, you can't (and shouldn't :P )

If you're exposing an API like that, you're pretty much stuck to it.

Reply:
I was asking because I wanted to see if in Scala things were different w.r.t. what I already knew. Phew! Nothing new here :D FYI, some languages *may do return type overloading*, but since almost everyone adheres to [CLS](https://msdn.microsoft.com/en-us/library/12a7a7h3(v=vs.100).aspx) they are not doing that *by design*.

Reply:
Are you sure? I knew the C# spec forbids overloading by return type.

Reply:
C# specs forbids overloading by return type, is exactly what I said.
What i was stressing out, is that it is like that not because of a [*technical reason*](https://stackoverflow.com/a/20705713), i.e. it was done like that *by design*.

Reply:
Ah yes, it's not technically impossible, it just makes method selection very hard, because you need to consider the context in which the method is called, and not just its invocation.